// @arkiv-network/sdk/dist/_cjs/actions/public/subscribeEntityEvents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.arkivABI = void 0;
exports.subscribeEntityEvents = subscribeEntityEvents;
const viem_1 = require("viem");
exports.arkivABI = (0, viem_1.parseAbi)([
    "event ArkivEntityCreated(uint256 indexed entityKey, address indexed ownerAddress, uint256 expirationBlock, uint256 cost)",
    "event ArkivEntityUpdated(uint256 indexed entityKey, address indexed ownerAddress, uint256 oldExpirationBlock, uint256 newExpirationBlock, uint256 cost)",
    "event ArkivEntityExpired(uint256 indexed entityKey, address indexed ownerAddress)",
    "event ArkivEntityDeleted(uint256 indexed entityKey, address indexed ownerAddress)",
    "event ArkivEntityBTLExtended(uint256 indexed entityKey, address indexed ownerAddress, uint256 oldExpirationBlock, uint256 newExpirationBlock, uint256 cost)",
]);
async function subscribeEntityEvents(client, { onError, onEntityCreated, onEntityUpdated, onEntityDeleted, onEntityExpired, onEntityExpiresInExtended, }, pollingInterval, fromBlock) {
    const unsubscribe = client.watchEvent({
        pollingInterval: pollingInterval ?? 1000,
        fromBlock: fromBlock ?? 0n,
        events: exports.arkivABI,
        onLogs: (logs) => {
            console.debug("logs from subscribeEntityEvents", logs);
            for (const log of logs) {
                const event = (0, viem_1.decodeEventLog)({
                    abi: exports.arkivABI,
                    topics: log.topics,
                    data: log.data,
                });
                console.debug("event from subscribeEntityEvents", event);
                switch (event.eventName) {
                    case "ArkivEntityCreated":
                        onEntityCreated?.({
                            entityKey: (0, viem_1.toHex)(event.args.entityKey, { size: 32 }),
                            owner: event.args.ownerAddress,
                            expirationBlock: Number(event.args.expirationBlock),
                            cost: event.args.cost,
                        });
                        break;
                    case "ArkivEntityUpdated":
                        onEntityUpdated?.({
                            entityKey: (0, viem_1.toHex)(event.args.entityKey, { size: 32 }),
                            owner: event.args.ownerAddress,
                            oldExpirationBlock: Number(event.args.oldExpirationBlock),
                            newExpirationBlock: Number(event.args.newExpirationBlock),
                            cost: event.args.cost,
                        });
                        break;
                    case "ArkivEntityDeleted":
                        onEntityDeleted?.({
                            entityKey: (0, viem_1.toHex)(event.args.entityKey, { size: 32 }),
                            owner: event.args.ownerAddress,
                        });
                        break;
                    case "ArkivEntityBTLExtended":
                        onEntityExpiresInExtended?.({
                            entityKey: (0, viem_1.toHex)(event.args.entityKey, { size: 32 }),
                            owner: event.args.ownerAddress,
                            oldExpirationBlock: Number(event.args.oldExpirationBlock),
                            newExpirationBlock: Number(event.args.newExpirationBlock),
                            cost: event.args.cost,
                        });
                        break;
                    case "ArkivEntityExpired":
                        onEntityExpired?.({
                            entityKey: (0, viem_1.toHex)(event.args.entityKey, { size: 32 }),
                            owner: event.args.ownerAddress,
                        });
                        break;
                    default:
                        console.warn("unknown event from subscribeEntityEvents", event);
                        break;
                }
            }
        },
        onError: (error) => {
            console.error("error from subscribeEntityEvents", error);
            onError?.(error);
        },
    });
    return unsubscribe;
}
