// @arkiv-network/sdk/dist/_esm/clients/decorators/arkivPublic.js
import { getBlockTiming } from "../../actions/public/getBlockTiming";
import { getEntity } from "../../actions/public/getEntity";
import { getEntityCount } from "../../actions/public/getEntityCount";
import { query } from "../../actions/public/query";
import { subscribeEntityEvents } from "../../actions/public/subscribeEntityEvents";
import { QueryBuilder } from "../../query/queryBuilder";
export function publicArkivActions(client) {
    return {
        getEntity: (key) => getEntity(client, key),
        query: (rawQuery) => query(client, rawQuery),
        buildQuery: () => new QueryBuilder(client),
        getBlockTiming: () => getBlockTiming(client),
        getEntityCount: () => getEntityCount(client),
        subscribeEntityEvents: ({ onError, onEntityCreated, onEntityUpdated, onEntityDeleted, onEntityExpired, onEntityExpiresInExtended, }, pollingInterval, fromBlock) => subscribeEntityEvents(client, {
            onError,
            onEntityCreated,
            onEntityUpdated,
            onEntityDeleted,
            onEntityExpired,
            onEntityExpiresInExtended,
        }, pollingInterval, fromBlock),
    };
}
