// @arkiv-network/sdk/dist/_cjs/index.js
"use strict";
// Export all public APIs
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWalletClient = exports.createPublicClient = exports.webSocket = exports.toRlp = exports.toHex = exports.toBytes = exports.http = void 0;
// Re-export commonly used viem types for convenience
var viem_1 = require("viem");
Object.defineProperty(exports, "http", { enumerable: true, get: function () { return viem_1.http; } });
Object.defineProperty(exports, "toBytes", { enumerable: true, get: function () { return viem_1.toBytes; } });
Object.defineProperty(exports, "toHex", { enumerable: true, get: function () { return viem_1.toHex; } });
Object.defineProperty(exports, "toRlp", { enumerable: true, get: function () { return viem_1.toRlp; } });
Object.defineProperty(exports, "webSocket", { enumerable: true, get: function () { return viem_1.webSocket; } });
var createPublicClient_1 = require("./clients/createPublicClient");
Object.defineProperty(exports, "createPublicClient", { enumerable: true, get: function () { return createPublicClient_1.createPublicClient; } });
var createWalletClient_1 = require("./clients/createWalletClient");
Object.defineProperty(exports, "createWalletClient", { enumerable: true, get: function () { return createWalletClient_1.createWalletClient; } });
