// @arkiv-network/sdk/dist/_esm/actions/wallet/mutateEntities.js
import { opsToTxData, sendArkivTransaction } from "../../utils/arkivTransactions";
function parseReceipt(receipt, params) {
    const createdEntities = [];
    const updatedEntities = [];
    const deletedEntities = [];
    const extendedEntities = [];
    const totalCreates = params.creates?.length ?? 0;
    const totalUpdates = params.updates?.length ?? 0;
    const totalDeletes = params.deletes?.length ?? 0;
    // iterate over all logs and parse the event
    // each odd log is the interested ones, logs goes in order like: creates, deleted, updates, extends
    for (let i = 0; i < receipt.logs.length; i++) {
        const log = receipt.logs[i];
        if (i % 2 === 1)
            continue;
        const index = i / 2;
        if (index < totalCreates) {
            createdEntities.push(log.topics[1]);
        }
        else if (index < totalCreates + totalDeletes) {
            deletedEntities.push(log.topics[1]);
        }
        else if (index < totalCreates + totalUpdates + totalDeletes) {
            updatedEntities.push(log.topics[1]);
        }
        else {
            extendedEntities.push(log.topics[1]);
        }
    }
    return { createdEntities, updatedEntities, deletedEntities, extendedEntities };
}
export async function mutateEntities(client, data, txParams) {
    if (!data.creates && !data.updates && !data.deletes && !data.extensions) {
        throw new Error("No operations to perform");
    }
    const txData = opsToTxData({
        creates: data.creates ?? [],
        updates: data.updates ?? [],
        deletes: data.deletes ?? [],
        extensions: data.extensions ?? [],
    });
    const receipt = await sendArkivTransaction(client, txData, txParams);
    console.debug("Receipt from mutateEntities", receipt);
    const { createdEntities, updatedEntities, deletedEntities, extendedEntities } = parseReceipt(receipt, data);
    return {
        txHash: receipt.transactionHash,
        createdEntities,
        updatedEntities,
        deletedEntities,
        extendedEntities,
    };
}
