// @arkiv-network/sdk/dist/_cjs/utils/arkivTransactions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.opsToTxData = opsToTxData;
exports.sendArkivTransaction = sendArkivTransaction;
const viem_1 = require("viem");
const consts_1 = require("../consts");
const compression_1 = require("./compression");
function opsToTxData({ creates, updates, deletes, extensions, ownershipChanges, }) {
    function formatAttributes(attribute) {
        return [(0, viem_1.toHex)(attribute.key), (0, viem_1.toHex)(attribute.value)];
    }
    const payload = [
        //creates
        (creates ?? []).map((item) => [
            (0, viem_1.toHex)(item.expiresIn / consts_1.BLOCK_TIME),
            (0, viem_1.toHex)(item.contentType),
            (0, viem_1.toHex)(item.payload),
            item.attributes
                .filter((attribute) => typeof attribute.value === "string")
                .map(formatAttributes),
            item.attributes
                .filter((attribute) => typeof attribute.value === "number")
                .map(formatAttributes),
        ]),
        //updates
        (updates ?? []).map((item) => [
            item.entityKey,
            (0, viem_1.toHex)(item.contentType),
            (0, viem_1.toHex)(item.expiresIn / consts_1.BLOCK_TIME),
            (0, viem_1.toHex)(item.payload),
            item.attributes
                .filter((attribute) => typeof attribute.value === "string")
                .map(formatAttributes),
            item.attributes
                .filter((attribute) => typeof attribute.value === "number")
                .map(formatAttributes),
        ]),
        //deletes
        (deletes ?? []).map((item) => item.entityKey),
        //extends
        (extensions ?? []).map((item) => [item.entityKey, (0, viem_1.toHex)(item.expiresIn / consts_1.BLOCK_TIME)]),
        //ownershipChanges TODO
        (ownershipChanges ?? []).map((item) => [item.entityKey, item.newOwner]),
    ];
    console.debug("txData to send as RLP", payload, payload.length);
    return (0, viem_1.toRlp)(payload);
}
async function sendArkivTransaction(client, data, txParams) {
    if (!client.account)
        throw new Error("Account required");
    const walletClient = client;
    console.debug("Sending transaction", {
        account: client.account,
        chain: client.chain,
        to: consts_1.ARKIV_ADDRESS,
        value: 0n,
        data,
        ...txParams,
    });
    const compressed = await (0, compression_1.compress)((0, viem_1.toBytes)(data));
    const txHash = await walletClient.sendTransaction({
        account: client.account,
        chain: client.chain,
        to: consts_1.ARKIV_ADDRESS,
        value: 0n,
        data: (0, viem_1.toHex)(compressed),
        ...txParams,
    });
    const receipt = await walletClient.waitForTransactionReceipt({ hash: txHash });
    console.debug("Tx receipt", receipt);
    return receipt;
}
