// @arkiv-network/sdk/dist/_cjs/clients/decorators/arkivPublic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.publicArkivActions = publicArkivActions;
const getBlockTiming_1 = require("../../actions/public/getBlockTiming");
const getEntity_1 = require("../../actions/public/getEntity");
const getEntityCount_1 = require("../../actions/public/getEntityCount");
const query_1 = require("../../actions/public/query");
const subscribeEntityEvents_1 = require("../../actions/public/subscribeEntityEvents");
const queryBuilder_1 = require("../../query/queryBuilder");
function publicArkivActions(client) {
    return {
        getEntity: (key) => (0, getEntity_1.getEntity)(client, key),
        query: (rawQuery) => (0, query_1.query)(client, rawQuery),
        buildQuery: () => new queryBuilder_1.QueryBuilder(client),
        getBlockTiming: () => (0, getBlockTiming_1.getBlockTiming)(client),
        getEntityCount: () => (0, getEntityCount_1.getEntityCount)(client),
        subscribeEntityEvents: ({ onError, onEntityCreated, onEntityUpdated, onEntityDeleted, onEntityExpired, onEntityExpiresInExtended, }, pollingInterval, fromBlock) => (0, subscribeEntityEvents_1.subscribeEntityEvents)(client, {
            onError,
            onEntityCreated,
            onEntityUpdated,
            onEntityDeleted,
            onEntityExpired,
            onEntityExpiresInExtended,
        }, pollingInterval, fromBlock),
    };
}
