// @arkiv-network/sdk/dist/_cjs/clients/decorators/arkivWallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.walletArkivActions = walletArkivActions;
const changeOwnership_1 = require("../../actions/wallet/changeOwnership");
const createEntity_1 = require("../../actions/wallet/createEntity");
const deleteEntity_1 = require("../../actions/wallet/deleteEntity");
const extendEntity_1 = require("../../actions/wallet/extendEntity");
const mutateEntities_1 = require("../../actions/wallet/mutateEntities");
const updateEntity_1 = require("../../actions/wallet/updateEntity");
function walletArkivActions(client) {
    return {
        createEntity: (data, txParams) => (0, createEntity_1.createEntity)(client, data, txParams),
        updateEntity: (data, txParams) => (0, updateEntity_1.updateEntity)(client, data, txParams),
        deleteEntity: (data, txParams) => (0, deleteEntity_1.deleteEntity)(client, data, txParams),
        extendEntity: (data, txParams) => (0, extendEntity_1.extendEntity)(client, data, txParams),
        changeOwnership: (data, txParams) => (0, changeOwnership_1.changeOwnership)(client, data, txParams),
        mutateEntities: (data, txParams) => (0, mutateEntities_1.mutateEntities)(client, data, txParams),
    };
}
